📖 [Docs] Document the granting of API access (#1407)
operator-framework/operator-controller: docs/concepts/how-to-grant-api-access.md

Reviewer: I think this file should probably go in the tutorials or How-to guide sections, since the content is a couple of procedures with some example configurations.

Author: Moved to how-to's in the b836ea5 and updated mkdocs as well

Reviewer: cool cool. I'll give it another quick pass in the AM. Thanks!

Administrators can define standard roles to control access to the API resources provided by installed operators. If the operator bundle does not provide default roles, you can create them yourself.

### Default Roles

- **View ClusterRole**: Grants read-only access to all custom resource objects of specified API resources across the cluster.
- **Edit ClusterRole**: Allows modifying all custom resource objects within the cluster.
- **Admin ClusterRole**: Provides full permissions (create, update, delete) over all custom resource objects for the specified API resources across the cluster.

### Example: Defining a Custom "View" ClusterRole

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: custom-resource-view
rules:
- apiGroups:
  - <your-api-group>
  resources:
  - <your-custom-resources>
  verbs:
  - get
  - list
  - watch
```

### Example: Defining a Custom "Edit" ClusterRole

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: custom-resource-edit
rules:
- apiGroups:
  - <your-api-group>
  resources:
  - <your-custom-resources>
  verbs:
  - get
  - list
  - watch
  - create
  - update

>   - patch

Reviewer 2: Add `delete`?

Author: yep, missed it.

kind: ClusterRoleBinding
metadata:
  name: custom-resource-view-binding
subjects:
- kind: User
  name: <username> # Or use Group for group-based binding
roleRef:
  kind: ClusterRole
  name: custom-resource-view
  apiGroup: rbac.authorization.k8s.io
```

This binding grants `<username>` read-only access to the custom resource across all namespaces.

#### Example: RoleBinding

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: custom-resource-edit-binding
  namespace: <namespace>
subjects:
- kind: User
  name: <username>
roleRef:
  kind: Role
  name: custom-resource-edit
  apiGroup: rbac.authorization.k8s.io
```

This RoleBinding restricts permissions to a specific namespace.

### Option 2: Extending Default Kubernetes Roles

To automatically extend existing Kubernetes roles (e.g., the default `view`, `edit`, and `admin` roles), you can add **aggregation labels** to your CRDs. This allows users who already have `view`, `edit`, or `admin` roles to interact with the custom resource without needing additional RoleBindings.

#### Example: Adding Aggregation Labels to a CRD

```yaml
apiVersion: apiextensions.k8s.io/v1

> kind: CustomResourceDefinition

Reviewer 2: I thought aggregation labels only worked on ClusterRoles?

metadata:
  name: <your-crd-name>
  labels:
    rbac.authorization.k8s.io/aggregate-to-view: "true"
    rbac.authorization.k8s.io/aggregate-to-edit: "true"
    rbac.authorization.k8s.io/aggregate-to-admin: "true"
spec:
  ...
```

When these labels are applied, users with existing RoleBindings to the default `view`, `edit`, or `admin` ClusterRoles automatically inherit permissions to interact with this CRD.

> **Source**: [Kubernetes RBAC Aggregation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings)

---

## 3. Finding API Groups and Resources Provided by a Bundle

To create appropriate RBAC policies, you need to know which API groups and resources are exposed by the installed operator. You can inspect the installed CRDs and resources by running:

```bash

> kubectl get crds

Reviewer 2: A user can use label selectors to find CRDs owned by a specific cluster extension

```
kubectl get crds -l 'olm.operatorframework.io/owner-kind=ClusterExtension,olm.operatorframework.io/owner-name=<clusterExtensionName>'
```

Reviewer 2: Even better, use `jq`:

```console
$ kubectl get crd -l 'olm.operatorframework.io/owner-kind=ClusterExtension,olm.operatorframework.io/owner-name=argocd' -o json | jq -r '.items[] | .spec.names.plural + "." + .spec.group'
applications.argoproj.io
applicationsets.argoproj.io
appprojects.argoproj.io
argocdexports.argoproj.io
argocds.argoproj.io
```

Reviewer 2: Might want to move this further up to the first step... "once you have installed the ClusterExtension, you'll likely want to setup RBAC for users to interact with it. The first step is to discover which resources are provided by the extension..."

Author: yep, adding

# Granting Users Access to API Resources in OLM

> When cluster extensions are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators hold full management access to these resources by default, whereas non-administrative users may lack sufficient permissions. Such users often need designated permissions to create, view, or edit these Custom Resources.

Reviewer: Nits
```suggestion
When cluster extensions are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators have full management access to these resources by default, whereas non-administrative users might lack sufficient permissions. Cluster administrators must create the needed permissions to create, view, or edit these Custom Resources for these users.
```

> When operators or cluster extensions are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators hold full management access to these resources by default, whereas non-administrative users may lack sufficient permissions. Such users often need designated permissions to create, view, or edit these Custom Resources.

Reviewer 3: ```suggestion
When ClusterExtension are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators hold full management access to these resources by default, whereas non-administrative users may lack sufficient permissions. Such users often need designated permissions to create, view, or edit these Custom Resources.
```

Reviewer 3: The meeting convey seems to keep ClusterExtension only.

Author: yep, changed all refs from `operator` to `cluster extension`.

# Granting Users Access to API Resources in OLM

> When operators or cluster extensions are installed through OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. By default, cluster administrators may have full access to manage these resources, but non-admin users often require specific permissions to create, view, or edit these custom resource objects.

Reviewer 3: ```suggestion
When operators or cluster extensions are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators hold full management access to these resources by default, whereas non-administrative users may lack sufficient permissions. Such users often need designated permissions to create, view, or edit these Custom Resources.
```

Maybe?

Author: yeah, reads better and "managed" is more correct. thanks!

> When operators or cluster extensions are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators hold full management access to these resources by default, whereas non-administrative users may lack sufficient permissions. Such users often need designated permissions to create, view, or edit these Custom Resources.

Reviewer 4: Nit: we might want to use  "cluster extension" instead of operator here and in other place because with OLMv1 we want go beyond operators and be more generic.

This is general comment for the doc: not only for this line.

Author: I think by mentioning right off the top that operator or cluster extension are interchangeable it's okay to use operator for brevity in the rest of the doc. Plus it brings along readers who already know OLM as manging operators. Additionally, it's still in the Operator Framework, so it helps still make that connection. Perhaps down the road we start to really root out the all the operator refs?

Reviewer 4: That's fine with me.

cc @michaelryanpeter just in case you have an opinion on the wording.

Reviewer 3: The meeting convey seems to keep ClusterExtension only.

> When operators or cluster extensions are managed via OLM, they often provide Custom Resource Definitions (CRDs) that expose new API resources. Typically, cluster administrators hold full management access to these resources by default, whereas non-administrative users may lack sufficient permissions. Such users often need designated permissions to create, view, or edit these Custom Resources.

Author: I took out operator mention entirely.

> OLM does not automatically manage RBAC for users to interact with CRDs from installed packages. It is recommended that cluster administrators handle RBAC (Role-Based Access Control) to grant appropriate permissions to users. This guide provides steps on how to manually configure RBAC, focusing on creating ClusterRoles and binding them to users or groups.

Reviewer 3: ```suggestion
OLM does **not** automatically configure or manage RBAC for users to interact with the APIs provided by installed packages. It is recommended that cluster administrators manage RBAC (Role-Based Access Control) to grant appropriate permissions to non-administrative users. This guide outlines the steps to manually configure RBAC, with a focus on creating ClusterRoles and binding them to specific users or groups.
```

just a few nit suggestions

---

## 1. Creating Default ClusterRoles for API/CRD Access

> Administrators can define standard roles to control access to the API resources provided by installed operators. If the operator bundle does not provide default roles, you can create them yourself.

Author: Striking mention of bundle, cuz yeah it is isn't already defined and doesn't add much

```bash
kubectl get crds
```

This will list all available CRDs, and you can inspect individual CRDs for their API groups:

```bash
kubectl get crd <crd-name> -o yaml
```

A user can use label selectors to find CRDs owned by a specific cluster extension:

```bash
kubectl get crds -l 'olm.operatorframework.io/owner-kind=ClusterExtension,olm.operatorframework.io/owner-name=<clusterExtensionName>'
```

---

## 2. Creating Default ClusterRoles for API/CRD Access

Administrators can define standard roles to control access to the API resources provided by installed operators. If the operator does not provide default roles, you can create them yourself.

### Default Roles

- **View ClusterRole**: Grants read-only access to all custom resource objects of specified API resources across the cluster.
- **Edit ClusterRole**: Allows modifying all custom resource objects within the cluster.

> - **Admin ClusterRole**: Provides full permissions (create, update, delete) over all custom resource objects for the specified API resources across the cluster.

Reviewer 3: Should we not have only two types: edit and view?
What is the difference between edit and admin?

> - **View ClusterRole**: Grants read-only access to all custom resource objects of specified API resources across the cluster.
> - **Edit ClusterRole**: Allows modifying all custom resource objects within the cluster.
> - **Admin ClusterRole**: Provides full permissions (create, update, delete) over all custom resource objects for the specified API resources across the cluster.

Reviewer 3: ```suggestion
- **View ClusterRole**: Grants read-only access to all custom resource objects of specified API resources across the cluster. This role is intended for users who need visibility into the resources without any permissions to modify them. It’s ideal for monitoring purposes and limited access viewing.
- **Edit ClusterRole**: Allows users to modify all custom resource objects within the cluster. This role enables users to create, update, and delete resources, making it suitable for team members who need to manage resources but should not control RBAC or manage permissions for others.
- **Admin ClusterRole**: Provides full permissions (create, update, delete) over all custom resource objects for the specified API resources across the cluster. In addition to resource management, it grants users the ability to modify roles and bindings within the cluster, allowing them to delegate specific permissions to other users or groups as needed.
```

> - **View ClusterRole**: Grants read-only access to all custom resource objects of specified API resources across the cluster.
> - **Edit ClusterRole**: Allows modifying all custom resource objects within the cluster.
> - **Admin ClusterRole**: Provides full permissions (create, update, delete) over all custom resource objects for the specified API resources across the cluster.

Reviewer 3: ```suggestion
- **View ClusterRole**: Grants read-only access to all custom resource objects of specified API resources across the cluster. This role is intended for users who need visibility into the resources without any permissions to modify them. It’s ideal for monitoring purposes and limited access viewing.
- **Edit ClusterRole**: Allows users to modify all custom resource objects within the cluster. This role enables users to create, update, and delete resources, making it suitable for team members who need to manage resources but should not control RBAC or manage permissions for others.
- **Admin ClusterRole**: Provides full permissions (create, update, delete) over all custom resource objects for the specified API resources across the cluster. In addition to resource management, it grants users the ability to modify roles and bindings within the cluster, allowing them to delegate specific permissions to other users or groups as needed.
```

What about clarify the intention of each as the differences between Edit vs Admin?

Author: yeah I think that's a good clarification.

### Example: Defining a Custom "View" ClusterRole

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: custom-resource-view
rules:
- apiGroups:
  - <your-api-group>
  resources:
  - <your-custom-resources>
  verbs:
  - get
  - list
  - watch
```

### Example: Defining a Custom "Edit" ClusterRole

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: custom-resource-edit
rules:
- apiGroups:
  - <your-api-group>
  resources:
  - <your-custom-resources>
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - patch
  - delete 
```

### Example: Defining a Custom "Admin" ClusterRole

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: custom-resource-admin
rules:
- apiGroups:
  - <your-api-group>
  resources:
  - <your-custom-resources>
  verbs:
  - '*'
```

> 

Reviewer 3: ```suggestion
**Note**: The '*' in verbs allows all actions on the specified resources, including RBAC management actions such as assigning roles and creating role bindings.
```
Wdyt?

Author: yeah, can't hurt to draw attention to the importance of granting '*"

#### Example: Adding Aggregation Labels to a ClusterRole

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: custom-resource-aggregated-view
  labels:
    rbac.authorization.k8s.io/aggregate-to-view: "true"
rules:
  - apiGroups:
      - <your-api-group>
    resources:
      - <your-custom-resource>
    verbs:
      - get
      - list
      - watch
```

You can create similar ClusterRoles for `edit` and `admin` with appropriate verbs (such as `create`, `update`, `delete` for `edit` and `admin`). By using aggregation labels, the permissions for the custom resources are added to the default roles.

> **Source**: [Kubernetes RBAC Aggregation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings)

---

## Notes

- OLM does not handle RBAC for users interacting with CRDs, so it's up to cluster administrators to configure these settings.

> - It is not recommended for operator bundles to include RBAC policies granting access to the operator's APIs because cluster administrators should maintain control over the permissions in their clusters.

Reviewer 3: Currently, in Kubebuilder, we scaffold only the edit and view RBAC roles, which allows projects to easily provide these roles to their administrators. (For example, see: [Kubebuilder RBAC configuration](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v4/config/rbac)).

_🙏 Your PR highlighted a potential improvement: we could enhance this scaffolding by adding admin roles and including aggregated role annotations. Thanks for bringing this to light—I'll raise an issue to address it._

❓ **On a related note, I have a question**: With the new OLMv1, we’re moving toward a more flexible packaging approach, where we could package various files, such as Helm charts, in the future. Am I right? Then:
- a) Should we recommend users include edit, view, and admin roles in their OLM bundles to allow administrative users more seamless management? 
- b) Is it currently possible or not to include these roles in my OLM bundle?

Author: 👍 on the issue add upstream.

idk on additional RBAC recommendation because of additional items such as Helm charts, etc. I don't think this doc needs to change for now. We can keep it in mind for once we offer Helm support, etc. Esp. if we see users getting into difficulty in using operators that don't give themselves much RBAC and install these types.

Reviewer 3: ```suggestion
- It is recommended that Operator authors include cluster administrative RBAC policies with their solution, as described above. This includes providing view, edit, and admin roles for each API, using the respective Kubernetes annotations. Doing so allows cluster administrators to more easily manage and grant access to the Operator’s APIs for non-administrative cluster users.
```

IHMO, **we should recommend that Operator authors include RBAC rules that allow admins to manage their solutions**. Imagine being a cluster admin with around 50 Operators installed. Would you want to manually configure RBAC rules for each one of the 50 solution—and then recheck them every time there’s an upgrade, since new APIs and permissions might be introduced or modified? Ensuring the correct RBAC rules are in place for management and access on an ongoing basis would create a poor and painful UX for cluster administrators.

Additionally, the guidance above implies the opposite to authors, effectively suggesting, "do not include these rules with your solution because it’s considered a bad practice," which doesn’t seem quite right.

Author: The recent commit adds language from the Slack convo. Thanks!

Author: resolved with new wording from the slack convo
